### PR TITLE
[Fix #7739] Allow methods raising NotImplementedError in `Lint/UnusedMethodArgument`

### LINKED ISSUE #7739: Allow unused variables when a class method only raises a NotImplementedError
## Is your feature request related to a problem? Please describe.

While the [style guide](https://github.com/rubocop-hq/ruby-style-guide#duck-typing) prefers duck typing over inheritance, many developers prefer inheritance for the sake of communicating structure to other developers. If a method both has arguments and raises a `NotImplementedError` – a common pattern for defining method signatures – the `Lint/UnusedMethodArgument` warning is triggered.

## Describe the solution you'd like

I would like functions with arguments that only contain `raise NotImplementedError` to be ignored by `Lint/UnusedMethodArgument`. Ideally, there would be a setting (maybe `IgnoreNotImplementedMethods`) that is true by default.

This should not trigger any warnings

```
class Example
  def process_thing(thing)
    raise NotImplementedError
  end
end
```

## Describe alternatives you've considered

Currently you can use the following work arounds

#### Leave a method empty, because `IgnoreEmptyMethods` is true by default
This is insufficient, because in many cases calling a method should immediately fail if it hasn't been implemented by a child class. Otherwise, bugs can appear far deeper in the execution path.

#### Prefix the variables with an underscore
This is insufficient, as it communicates that the prefixed argument(s) are optional. They may not be.

#### Don't include arguments in the signature
This is insufficient, as it fails to communicate the arguments child classes should expect for a method.

### PR DESCRIPTION
**Pull request by tejasbubane**
Closes #7739

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/lint/unused_method_argument.rb [hunk 1]
```diff
@@ -38,9 +38,25 @@ module Lint
...
+      #   # bad
+      #   def do_something(unused)
+      #     raise NotImplementedError
+      #   end
+      #
       class UnusedMethodArgument < Cop
         include UnusedArgument
 
+        def_node_matcher :raise_not_implemented?, <<~PATTERN
+          (send nil? :raise (const nil? :NotImplementedError))
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/unused_method_argument.rb]: You should also cover `fail`.

File: config/default.yml [hunk 2]
```diff
@@ -1739,6 +1739,7 @@ Lint/UnusedMethodArgument:
   VersionChanged: '0.35'
   AllowUnusedKeywordArguments: false
   IgnoreEmptyMethods: true
+  IgnoreNotImplementedMethods: true
```
Comment (bbatsov) [inline config/default.yml]: Version changed needs to be updated.

**Comment by bbatsov** [comment]
Apart from my small remarks I'm fine with the proposed changes.

File: config/default.yml (hunk 2 above)
Comment (tejasbubane) [reply config/default.yml]: Updated.

File: lib/rubocop/cop/lint/unused_method_argument.rb (hunk 1 above)
Comment (tejasbubane) [reply lib/rubocop/cop/lint/unused_method_argument.rb]: Handled.

**Comment by bbatsov** [comment]
Thanks!

**Comment by sandstrom** [comment]
@tejasbubane Awesome addition! 

What do you think about [Lint/UnreachableCode](https://github.com/rubocop-hq/rubocop/blob/master/lib/rubocop/cop/lint/unreachable_code.rb) having this check too?

Basically such that it would allow:

```ruby
def my_method
  raise NotImplementedError
  
  1 + 1
end
```

**Pull request merged**